sites/game-client/ui/context-menu.mjs

```diff
@@ -0,0 +1,157 @@
+// Copyright 2021 HITCON Online Contributors
+// SPDX-License-Identifier: BSD-2-Clause
+
+const CONTEXT_MENU_OTHER_DIV = 'context-menu-other';
+const CONTEXT_MENU_LIST_OTHER_UL = 'context-menu-list-other';
+const CONTEXT_MENU_SELF_DIV = 'context-menu-self';
+const CONTEXT_MENU_LIST_SELF_UL = 'context-menu-list-self';
+
+const mapCellSize = 32;
+
+/**
+ * Maintain the state of the context menu for players.
+*/
+class ContextMenu {
+    /**
+     *
+     */
+    constructor(gameState, mapRenderer) {
+        this.gameState = gameState;
+        this.mapRenderer = mapRenderer;
+        this.othersMenu = new Map();
+        this.selfMenu = new Map();
+
+        this.focusedPlayer = null;
+        this.playerInfo = null;
+
+        // bind the contextmenu and click event.
+        document.getElementById('mapcanvas').addEventListener('contextmenu', this.onContextMenu.bind(this));
+        document.body.addEventListener('click', this.hideMenu);
+
+        // Get current player
+        window.addEventListener('gameStart', (d) => {
+            this.playerInfo = d.detail.gameClient.playerInfo;
+        });
+    }
+
+    /**
+     * Called when user clicked on the player.
+     * @param {Event} e event
+     */
+    onContextMenu(e) {
+        e.preventDefault();
+
+        // Check if the user actually clicked on a player.
+        if (!this.findFocusedUser(e.pageX, e.pageY)) {
+            return;
+        }
+
+        // Identify whether the menu is for the user itself or other players.
+        // Stop if the menu is empty.
+        let div;
+        if (this.focusedPlayer.playerID === this.playerInfo.playerID && this.selfMenu.size > 0) {
+            div = CONTEXT_MENU_SELF_DIV;
+        } else if (this.focusedPlayer.playerID !== this.playerInfo.playerID && this.othersMenu.size > 0) {
+            div = CONTEXT_MENU_OTHER_DIV;
+        } else {
+            this.focusedPlayer = null;
+            return;
+        }
```

Comment:
Optional: I would generally recommend splitting the logic between opening the context menu and triggering it from events from the canvas. The reason is because there may be multiple causes for triggering the context menu, such as people right clicking on the username in the chat overlay. This change can be delayed to later when we add the trigger for context menu in chat overlay.